app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js

```diff
@@ -163,26 +170,90 @@ export async function switchChain(
     setActiveNetwork,
     endApprovalFlow,
     getCaveat,
-    requestPermittedChainsPermission,
-    grantPermittedChainsPermissionIncremental,
+    requestPermissionApprovalForOrigin,
+    updateCaveat,
+    grantPermissions,
   },
 ) {
   try {
-    const { value: permissionedChainIds } =
-      getCaveat({
-        target: PermissionNames.permittedChains,
-        caveatType: CaveatTypes.restrictNetworkSwitching,
-      }) ?? {};
+    const caip25Caveat = getCaveat({
+      target: Caip25EndowmentPermissionName,
+      caveatType: Caip25CaveatType,
+    });
 
-    if (
-      permissionedChainIds === undefined ||
-      !permissionedChainIds.includes(chainId)
-    ) {
-      if (isAddFlow) {
-        await grantPermittedChainsPermissionIncremental([chainId]);
-      } else {
-        await requestPermittedChainsPermission([chainId]);
+    if (caip25Caveat) {
+      const ethChainIds = getPermittedEthChainIds(caip25Caveat.value);
+
+      if (!ethChainIds.includes(chainId)) {
+        if (caip25Caveat.value.isMultichainOrigin) {
+          return end(
+            ethErrors.provider.unauthorized(
+              `Cannot switch to or add permissions for chainId '${chainId}' because permissions were granted over the Multichain API.`,
+            ),
+          );
+        }
+
+        // TODO: This behavior may have deviated from the original permittedChains add chain behavior
+        // Verify that this helper behaves as expected
```

Comment:
TODO

Reply:
also reminder: 

> revoking permittedChains revokes the entire CAIP-25 permission. Should it do this, or just remove all eip155:x scopes (leaving wallet:eip155 intact so that we have somewhere to store eth accounts)?